segmentstore/server/src/main/java/io/pravega/segmentstore/server/reading/ContainerReadIndex.java

```diff
@@ -189,8 +190,24 @@ public void triggerFutureReads(Collection<Long> streamSegmentIds) {
                         missingIds.add(segmentId);
                     }
                 }
-            } else {
+                continue;
+            }
+
+            try {
                 index.triggerFutureReads();
+            } catch (ObjectClosedException ex) {
+                // It is possible that between the time we got the pointer to the StreamSegmentReadIndex and when we got
+                // to invoking triggerFutureReads, the StreamSegmentReadIndex has already been closed. If this is the case,
+                // ignore the error.
+                // This is possible in the following scenario: for a Transaction, we have an Append, followed by a Merge;
+                // the Append makes this index eligible for triggering future reads, and the Merge (once committed to Storage)
+                // will close it. If the StorageWriter is sufficiently fast in comparison to the OperationProcessor callbacks
+                // (which could be the case for in-memory unit tests), it may trigger this condition.
+                if (getIndex(segmentId) == null) {
```

Comment:
It would, but only if it got closed before the call to`getIndex` above. The `StorageWriter` is an async process which may invoke methods on this class from different threads. It is quite possible that the index is closed after `getIndex` and before we invoke `triggerFutureReads` on it - which causes this error.

`getIndex` will return the index if it is registered. However when we close the index (see `closeIndex`) we atomically close it and unregister it, so it will never be the case that we get a reference to a closed index.